### PR TITLE
big change! upgrade zktrie circuit

### PR DESCRIPTION
**Pull request by z2trillion**
"updates in mpt table proven in mpt circuit" when `make test-heavy`

### CODE REVIEW

File: zkevm-circuits/src/mpt_circuit.rs [hunk 1]
```diff
@@ -7,51 +8,66 @@ use crate::{
...
-                poseidon_table.hash_id,
-                poseidon_table.input0,
-                poseidon_table.input1,
-                poseidon_table.control,
-                poseidon_table.heading_mark,
-            ],
-        );
-        let mpt_table = (
-            mpt_table.q_enable,
+        let _mpt_table_inp = //(
```
Comment (lispc) [inline zkevm-circuits/src/mpt_circuit.rs]: is this correct? @z2trillion 

state circuit uses the mpt_table allocated outside, while mpt circuit does not use this mpt table at all.  So i think there must be missing some constraints?

File: zkevm-circuits/src/witness/mpt.rs [hunk 2]
```diff
@@ -345,3 +382,1168 @@ fn value_prev(row: &Rw) -> Word {
...
+            original_rws: Default::default(),
+        });
+
+        updates.fill_state_roots(&ZktrieState::default());
+        println!(
+            "{}",
+            serde_json::to_string_pretty(&updates.smt_traces.last().unwrap()).unwrap()
+        );
+
+        panic!();
```
Comment (lispc) [inline zkevm-circuits/src/witness/mpt.rs]: ignore the test or fix the panic

File: zkevm-circuits/src/witness/mpt.rs [hunk 3]
```diff
@@ -23,16 +27,29 @@ pub struct MptUpdate {
...
+                }
+            }
+            Key::Account { field_tag, .. } => {
+                dbg!(field_tag);
+                if matches!(field_tag, AccountFieldTag::CodeHash)
+                    && self.old_value.is_zero()
+                    && self.new_value.is_zero()
+                {
+                    dbg!(field_tag, self.old_value, self.new_value);
+                    dbg!(matches!(field_tag, AccountFieldTag::CodeHash));
```
Comment (lispc) [inline zkevm-circuits/src/witness/mpt.rs]: remove these dbg?

File: zkevm-circuits/src/mpt_circuit.rs (hunk 1 above)
Comment (z2trillion) [reply zkevm-circuits/src/mpt_circuit.rs]: yes, these constraints are currently still missing.

Comment (z2trillion) [reply zkevm-circuits/src/mpt_circuit.rs]: https://github.com/scroll-tech/zkevm-circuits/pull/551  along with https://github.com/scroll-tech/mpt-circuit/pull/47 will fix this.

**Comment by z2trillion** [comment]
https://github.com/scroll-tech/mpt-circuit/pull/48 should fix the one failing test regarding the super circuit degree.

File: zkevm-circuits/src/mpt_circuit.rs [hunk 4]
```diff
@@ -7,51 +8,66 @@ use crate::{
...
+    }
+}
+
+/// Circuit wrapped with mpt table data
+#[derive(Clone, Debug, Default)]
+pub struct MptCircuit<F: Field> {
+    row_limit: usize,
+    proofs: Vec<Proof>,
+    mpt_updates: witness::MptUpdates,
+    _phantom: std::marker::PhantomData<F>,
```
Comment (z2trillion) [inline zkevm-circuits/src/mpt_circuit.rs:38]: I don't think there's a need for a phantom data marker?

**Review by z2trillion — COMMENTED** [review]
This looks good. I think we can remove the phantom data from MptCircuit though?

**Comment by lispc** [comment]
i will use this branch to run `make mock/make test-super-trace/make test-agg` in scroll-zkevm

**Comment by lispc** [comment]
oh now the CI(test-heavy, mainly super circuit) runs so long...


seems no good, i will cancel CI

**Comment by lispc** [comment]
i tried `cargo test --release --features scroll --all --exclude integration-tests --exclude circuit-benchmarks serial_test_super_circuit_1tx_1max_tx -- --ignored`

there are 3 types of failed lookups:

  Lookup 'account mpt key = h(address_high, address_low)' 
  Lookup 'poseidon hash correct for old common path' 
  Lookup 'sibling = h(1, key)' 

i guess they are all related to (0,0) => 0 in poseidon...


@z2trillion can have some fixes here after you return to work?

**Comment by noel2004** [comment]
> i tried `cargo test --release --features scroll --all --exclude integration-tests --exclude circuit-benchmarks serial_test_super_circuit_1tx_1max_tx -- --ignored`
> 
> there are 3 types of failed lookups:
> 
> Lookup 'account mpt key = h(address_high, address_low)' Lookup 'poseidon hash correct for old common path' Lookup 'sibling = h(1, key)'
> 
> i guess they are all related to (0,0) => 0 in poseidon...
> 
> @z2trillion can have some fixes here after you return to work?

Poseidon circuit used `max_evm_row`, which is not set in unittest. It is one of the causes of test failure (another is the [lookup 'updates in mpt table proven in mpt circuit'](https://github.com/scroll-tech/zkevm-circuits/blob/994f6e682a026bca3b982bb5625c90211da768a7/zkevm-circuits/src/mpt_circuit.rs#L72) failed). https://github.com/scroll-tech/zkevm-circuits/pull/548/commits/aedad4da13ff169dffd520c8dabb735d78b7c796 has fixed this. @z2trillion

**Comment by lispc** [comment]
a lot of 'updates in mpt table proven in mpt circuit' when `make test-heavy`

**Comment by noel2004** [comment]
https://github.com/scroll-tech/zkevm-circuits/blob/b480692a22c5be06b72dbcdf2be2cefe8e222342/zkevm-circuits/src/witness/mpt.rs#L1394 here insert a mpt update entry which has written a new value into an empty account address. 

It cause a panic in a sanity check and made two test `witness::mpt::test::empty_account_empty_storage_proof` and `witness::mpt::test::empty_account_storage_write` not pass.

Wondering why we need such a mpt update entry here, I guess this is impossible for any valid circuit. @z2trillion

**Comment by z2trillion** [comment]
> https://github.com/scroll-tech/zkevm-circuits/blob/b480692a22c5be06b72dbcdf2be2cefe8e222342/zkevm-circuits/src/witness/mpt.rs#L1394
> here insert a mpt update entry which has written a new value into an empty account address.
> 
> It cause a panic in a sanity check and made two test `witness::mpt::test::empty_account_empty_storage_proof` and `witness::mpt::test::empty_account_storage_write` not pass.
> 
> Wondering why we need such a mpt update entry here, I guess this is impossible for any valid circuit. @z2trillion

These aren't really testing anything. They explicitly panic at the end so that the debug statements are printed out. The resulting json's are using as test traces in the mpt circuit repo.

It is possible to need an empty storage proof for for an empty account. An example of this is a revert(account creation tx + storage write). In the state circuit, we will need to prove that the initial and final values for storage key for an empty account are both 0. This mpt circuit in the v0.4 branch handles this case, which is covered by the test `empty_account_empty_storage`.

**Comment by noel2004** [comment]
> > https://github.com/scroll-tech/zkevm-circuits/blob/b480692a22c5be06b72dbcdf2be2cefe8e222342/zkevm-circuits/src/witness/mpt.rs#L1394
> > 
> > here insert a mpt update entry which has written a new value into an empty account address.
> > It cause a panic in a sanity check and made two test `witness::mpt::test::empty_account_empty_storage_proof` and `witness::mpt::test::empty_account_storage_write` not pass.
> > Wondering why we need such a mpt update entry here, I guess this is impossible for any valid circuit. @z2trillion
> 
> These aren't really testing anything. They explicitly panic at the end so that the debug statements are printed out. The resulting json's are using as test traces in the mpt circuit repo.
> 
> It is possible to need an empty storage proof for for an empty account. An example of this is a revert(account creation tx + storage write). In the state circuit, we will need to prove that the initial and final values for storage key for an empty account are both 0. This mpt circuit in the v0.4 branch handles this case, which is covered by the test `empty_account_empty_storage`.

Agree. and empty storage proof (so both `value_before` and `value_after` is 0) would not trigger panic here.

**Comment by lispc** [comment]
finally .... i will merge this once super proof can be generated inside scroll-zkevm. Probably within 1h.

**Pull request merged**